stablehlo/reference/InterpreterApi.cpp

```diff
@@ -0,0 +1,166 @@
+limitations under the License.
+==============================================================================*/
+#include "stablehlo/reference/InterpreterApi.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/SourceMgr.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "stablehlo/reference/Errors.h"
+#include "stablehlo/reference/InterpreterOps.h"
+#include "stablehlo/reference/NumPy.h"
+#include "stablehlo/reference/Ops.h"
+
+namespace mlir {
+namespace stablehlo {
+namespace {
+
+llvm::Error wrapStatus(llvm::Error status, llvm::StringRef funcName,
+                       llvm::StringRef fallbackName) {
+  if (status)
+    return stablehlo::invalidArgument(
+        "Error evaluating function: %s. \n\tFallback for %s failed: %s",
+        funcName.data(), fallbackName.data(),
+        toString(std::move(status)).c_str());
+  return llvm::Error::success();
+}
+
+}  // namespace
+
+llvm::Error InterpreterFallback::operator()(Operation &op, Process *process,
+                                            Scope &scope) {
+  llvm::StringRef funcName = currentFcn.getSymName();
+
+  if (auto probeOp = dyn_cast<stablehlo::interpreter::ProbeOp>(op)) {
+    auto input =
+        stablehlo::InterpreterValue(scope.findTensor(probeOp.getOperand()));
+    auto status = stablehlo::interpreter::evalProbeOp(
+        input, probeOp.getProbeId(), config->probeInstrumentationDir,
+        instrumentedTensors);
+    scope.add(probeOp.getResult(), input);
+    return wrapStatus(std::move(status), funcName, "interpreter.probe");
+  }
+
+  if (auto runParallelOp =
+          dyn_cast<stablehlo::interpreter::RunParallelOp>(op)) {
+    auto runtimeOperands = scope.find(runParallelOp.getInputs());
+    std::queue<StringAttr> infeed;
+    if (auto infeedAttr = runParallelOp.getInfeed())
+      for (auto &value : infeedAttr->getValue())
+        infeed.push(value.cast<FlatSymbolRefAttr>().getAttr());
+
+    SmallVector<SmallVector<StringAttr>> programs(
+        runParallelOp.getPrograms().size());
+    for (auto [i, replica] : llvm::enumerate(runParallelOp.getPrograms()))
+      for (auto &program : replica.cast<ArrayAttr>())
+        programs[i].push_back(program.cast<FlatSymbolRefAttr>().getAttr());
+
+    SymbolTable symbolTable{op.getParentOfType<ModuleOp>()};
+    auto results = stablehlo::interpreter::evalRunParallelOp(
+        runtimeOperands, infeed, programs, symbolTable);
+    scope.add(runParallelOp.getResults(), results);
+    return wrapStatus(llvm::Error::success(), funcName,
+                      "interpreter.run_parallel");
+  }
+
+  return handleOp(op, process, scope);
+}
+
+llvm::Error InterpreterFallback::handleOp(Operation &op, Process *process,
```

Comment:
I don't think we need this abstracted out yet. Since we don't handle them in multiple ways, let's keep the error message inlined?

Reply:
The initial idea behind splitting this was that if/when user specify a custom fallback, they need not worry about explicitly inheriting or calling the interpreter ops fallback. That is, interpreter ops would always have a kernel defined.

Were you thinking of having the interpreter op kernels be optional in a singular overridable callback?